server/etcdserver/apply/uber_applier.go

```diff
@@ -0,0 +1,236 @@
+	"go.etcd.io/etcd/server/v3/etcdserver/api/membership"
+	"go.etcd.io/etcd/server/v3/etcdserver/api/v3alarm"
+	"go.etcd.io/etcd/server/v3/etcdserver/cindex"
+	"go.etcd.io/etcd/server/v3/etcdserver/txn"
+	"go.etcd.io/etcd/server/v3/lease"
+	"go.etcd.io/etcd/server/v3/storage/backend"
+	"go.etcd.io/etcd/server/v3/storage/mvcc"
+	"go.uber.org/zap"
+)
+
+type UberApplier struct {
+	lg *zap.Logger
+
+	alarmStore           *v3alarm.AlarmStore
+	warningApplyDuration time.Duration
+
+	// This is the applier that is taking in consideration current alarms
+	applyV3 applierV3
+
+	// This is the applier used for wrapping when alarms change
+	applyV3base applierV3
+}
+
+func NewUberApplier(
+	lg *zap.Logger,
+	be backend.Backend,
+	kv mvcc.KV,
+	alarmStore *v3alarm.AlarmStore,
+	authStore auth.AuthStore,
+	lessor lease.Lessor,
+	cluster *membership.RaftCluster,
+	raftStatus RaftStatusGetter,
+	snapshotServer SnapshotServer,
+	consistentIndex cindex.ConsistentIndexer,
+	warningApplyDuration time.Duration,
+	txnModeWriteWithSharedBuffer bool,
+	quotaBackendBytesCfg int64) *UberApplier {
+	applyV3base_ := newApplierV3(lg, be, kv, alarmStore, authStore, lessor, cluster, raftStatus, snapshotServer, consistentIndex, txnModeWriteWithSharedBuffer, quotaBackendBytesCfg)
+
+	ua := &UberApplier{
+		lg:                   lg,
+		alarmStore:           alarmStore,
+		warningApplyDuration: warningApplyDuration,
+		applyV3:              applyV3base_,
+		applyV3base:          applyV3base_,
+	}
+	ua.RestoreAlarms()
+	return ua
+}
+
+func newApplierV3(
+	lg *zap.Logger,
+	be backend.Backend,
+	kv mvcc.KV,
+	alarmStore *v3alarm.AlarmStore,
+	authStore auth.AuthStore,
+	lessor lease.Lessor,
+	cluster *membership.RaftCluster,
+	raftStatus RaftStatusGetter,
+	snapshotServer SnapshotServer,
+	consistentIndex cindex.ConsistentIndexer,
+	txnModeWriteWithSharedBuffer bool,
+	quotaBackendBytesCfg int64) applierV3 {
+	applierBackend := newApplierV3Backend(lg, kv, alarmStore, authStore, lessor, cluster, raftStatus, snapshotServer, consistentIndex, txnModeWriteWithSharedBuffer)
+	return newAuthApplierV3(
+		authStore,
+		newQuotaApplierV3(lg, quotaBackendBytesCfg, be, applierBackend),
+		lessor,
+	)
+}
+
+func (a *UberApplier) RestoreAlarms() {
+	noSpaceAlarms := len(a.alarmStore.Get(pb.AlarmType_NOSPACE)) > 0
+	corruptAlarms := len(a.alarmStore.Get(pb.AlarmType_CORRUPT)) > 0
+	a.applyV3 = a.applyV3base
+	if noSpaceAlarms {
+		a.applyV3 = newApplierV3Capped(a.applyV3)
+	}
+	if corruptAlarms {
+		a.applyV3 = newApplierV3Corrupt(a.applyV3)
+	}
+}
+
+func (a *UberApplier) Apply(r *pb.InternalRaftRequest, shouldApplyV3 membership.ShouldApplyV3) *ApplyResult {
+	// We first execute chain of WrapApply across all objects (e.g. CorruptApplier -> CappedApplier -> Auth -> Quota -> Backend),
+	// than dispatch(), than individual methods wrappers CorruptApplier.Put(CappedApplier.Put(... BackendApplier.Put())))
+	return a.applyV3.WrapApply(context.TODO(), r, shouldApplyV3, a.dispatch)
```

Comment:
Why do yo introduce another layer of encapsulation/dispatch? I'd suggest to do it in the following PR, so that we can keep this PR as simple/clean as possible.

Reply:
It's not another layer, it's just making an explicit what was subtle & hidden so far. 

Please see this comment: 
https://github.com/etcd-io/etcd/blob/d849269dab5c07c1259de4441809d41135e7d0d8/server/etcdserver/apply.go#L166-L170

The call down the hierarchy was first traversing the recursive 'Apply`call structure, to get translated to a specific method, to traverse the overrides of the specific methods down the hierarchy. 

I broke the 'cyclic' dependency without changing the call hierarchy.

Reply:
1. Renamed - but to be honest I don't think it 'less' misleading.  
The function does not full 'Apply' it does only provides generic pre & post apply logic... and the actual application is delegated to the 'applyFunc' provided as a parameter.  
I have no better idea how to get rid of 'applyFunc'. 

Intuitively such stack of executions would be IMHO the most expected:
```
uberApplier.Apply:
  dispatcher()
    authApplierV3.Put
       authApplierV3.Pre
       quotaApplierV3.Put
           quotaApplierV3.Pre
           applierV3backend.Put
             applierV3backend.Pre
             (actual application)
             applierV3backend.Post
           quotaApplierV3.Post   
       authApplierV3.Post
 ```
 
 But I don't see how to implement it generically in go... in not error-prone way. 
 
 2. I unexported `ApplyFunc` type as it's indeed a local. 
    All the interfaces and implementations are 'private' (i.e. lower case names). 
    I have doubts about lower-case methods within the interfaces and classed. 
    Maybe I'm too biased in Java style of thinking, but in Java the interface methods are always public...
    but the interface itself can be package local.   I think that lower case methods of class are class private implementation detail, while the Capital methods are expected to be called by other classes within the same packages. I know it's not the golang specification... but also I don't see here risk that anyone would use the methods from outside of the package... as only exported think is `UberApplier` with a very narrow interface.

Reply:
This PR is already too big, so Let's discuss & address this in a separate session. The naming and structure looks much better now.